### PR TITLE
Aka users

### PR DESCRIPTION
**Pull request by Alihan26**
# Pull Requests Checklist

Assignee:

- \[x\] code commented?
- \[x\] tests added / passed?
- \[x\] are tests covering all cases?
- \[x\] documentation added / updated?
- \[x\] pre-commit checks passed?
- \[x\] pipelines passed?

Reviewer:

- \[x\] are there any code smells?
- \[ \] did pipeline pass?
- \[ \] is the code well documented?
- \[x\] are the tests passing and covering all cases?

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs24/service/UserService.java [hunk 1]
```diff
@@ -50,6 +50,47 @@ public User createUser(User newUser) {
...
+    log.debug("Deleted User: {}", existingUser);
+  }
+
+  // `checkInput` method in `UserService` checks the input fields of the user for null values.
+  // If any of them are null, it throws a `BAD_REQUEST` exception
+  public void checkInput(User userInput) {
+    if (userInput.getName() == null || userInput.getUsername() == null
+        || userInput.getPassword() == null) {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Input fields can't be null");
+    }
```
Comment (5v3nn) [inline src/main/java/ch/uzh/ifi/hase/soprafs24/service/UserService.java:91]: what this?

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:1]: please add other edge case tests. not just happy path testing

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java [hunk 2]
```diff
@@ -44,4 +44,25 @@ public UserGetDTO createUser(@RequestBody UserPostDTO userPostDTO) {
     // convert internal representation of user back to API
     return DTOMapper.INSTANCE.convertEntityToUserGetDTO(createdUser);
   }
+
+  @PutMapping("/users/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public UserGetDTO updateUser(@PathVariable Long id, @RequestBody UserPostDTO userPostDTO) {
+    // convert API user to internal representation
```
Comment (5v3nn) [inline src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:52]: add auth check if auth token is user's token

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java [hunk 3]
```diff
@@ -44,4 +44,25 @@ public UserGetDTO createUser(@RequestBody UserPostDTO userPostDTO) {
...
+    User updatedUser = userService.updateUser(userInput);
+
+    // convert internal representation of user back to API
+    return DTOMapper.INSTANCE.convertEntityToUserGetDTO(updatedUser);
+  }
+
+  @DeleteMapping("/users/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  public void deleteUser(@PathVariable Long id) {
+    userService.deleteUser(id);
```
Comment (5v3nn) [inline src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:66]: add auth check if auth token is user's token

**Review by 5v3nn — CHANGES_REQUESTED** [review]
please fix

File: src/main/java/ch/uzh/ifi/hase/soprafs24/service/UserService.java (hunk 1 above)
Comment (Alihan26) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/service/UserService.java]: This is a method, which I used to implement my methods in UserController, as it was in the issue 52, yet I changed the implementation, didn't need this method anymore, and forgot to delete it, now it's gone

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java
Comment (Alihan26) [reply src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:1]: Yes you're right, for PUT, I have created a test case where it fails to find the user to be modified

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 2 above)
Comment (Alihan26) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:56]: This has been modified now using the code line below, can you check if it's okay?

```
// Check if the token is valid and belongs to the user
    authorizationService.isAuthorized(token);
```

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 3 above)
Comment (Alihan26) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:73]: the code line below has been inserted to check for the auth token, if it's indeed the user's token

```
// Check if the token is valid and belongs to the user
    authorizationService.isAuthorized(token);
```

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 2 above)
Comment (5v3nn) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:62]: this now only checks if the token is in the db. we should also check if the token is the user's token. use this instead:
```java
authorizationService.isAuthorized(token, id);`
```

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 3 above)
Comment (5v3nn) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:79]: same as above

**Review by 5v3nn — CHANGES_REQUESTED** [review]

**Comment by 5v3nn** [comment]
@Alihan26 please add service integration tests:

for both update and delete
- [ ] if user not found
- [ ] update: if not valid input
- [ ] sucessful

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java [hunk 4]
```diff
@@ -101,6 +106,140 @@ public void createUser_duplicateUsername_throwsError() throws Exception {
...
+
+  // Alihan: Delete user (happy path)
+  @Test
+  public void deleteUser_validInput_userDeleteda() throws Exception {
+    // given
+    User user = new User();
+    user.setUserId(1L);
+    user.setName("Test User");
+    user.setUsername("testUsername");
+
```
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:198]: mock isAuthorized (because unit tests)

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java [hunk 5]
```diff
@@ -101,6 +106,140 @@ public void createUser_duplicateUsername_throwsError() throws Exception {
...
+            -> assertTrue(result.getResolvedException().getMessage().contains("User not found")));
+  }
+
+  // Alihan: Update User; Test case for PUT method where user tries to update with invalid input
+  @Test
+  public void updateUser_invalidInput_throwsError() throws Exception {
+    UserPostDTO userPostDTO = new UserPostDTO();
+    userPostDTO.setName("");
+    userPostDTO.setUsername("");
+
```
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:172]: mock isAuthorized (because unit tests)

File: src/test/java/ch/uzh/ifi/hase/soprafs24/service/UserServiceIntegrationTest.java [hunk 6]
```diff
@@ -105,4 +105,57 @@ public void createUser_duplicateName_successful() {
...
+  public void deleteUser_existingUser_success() {
+    User testUser = new User();
+    testUser.setName("testName");
+    testUser.setUsername("testUsername");
+    testUser.setPassword("1234");
+
+    User createdUser = userService.createUser(testUser);
+    userService.deleteUser(createdUser.getUserId());
+
+    assertNull(userRepository.findById(createdUser.getUserId()).orElse(null));
```
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/service/UserServiceIntegrationTest.java:153]: should not have `.orElse(null)`, because we want to find user

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java [hunk 7]
```diff
@@ -101,6 +106,140 @@ public void createUser_duplicateUsername_throwsError() throws Exception {
...
+    // given
+    User user = new User();
+    user.setUserId(1L);
+    user.setName("Test User Updated");
+    user.setUsername("testUsernameUpdated");
+
+    UserPostDTO userPostDTO = new UserPostDTO();
+    userPostDTO.setName("Test User Updated");
+    userPostDTO.setUsername("testUsernameUpdated");
+
```
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:121]: mock is authorized (becuase unit tests)

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java [hunk 8]
```diff
@@ -101,6 +106,140 @@ public void createUser_duplicateUsername_throwsError() throws Exception {
...
+  }
+
+  // Alihan: Update User; Test case for PUT method where user doesn't exist
+  @Test
+  public void updateUser_nonExistingUser_throwsError() throws Exception {
+    // given
+    UserPostDTO userPostDTO = new UserPostDTO();
+    userPostDTO.setName("Test User Updated");
+    userPostDTO.setUsername("testUsernameUpdated");
+
```
Comment (5v3nn) [inline src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:146]: mock is authorized (becuase unit tests)

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 3 above)
Comment (5v3nn) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:79]: ok

File: src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java (hunk 2 above)
Comment (5v3nn) [reply src/main/java/ch/uzh/ifi/hase/soprafs24/controller/UserController.java:62]: ok

**Review by 5v3nn — CHANGES_REQUESTED** [review]
please fix

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java (hunk 8 above)
Comment (5v3nn) [reply src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:156]: done

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java (hunk 7 above)
Comment (5v3nn) [reply src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:126]: done

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java (hunk 5 above)
Comment (5v3nn) [reply src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:206]: done

File: src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java (hunk 4 above)
Comment (5v3nn) [reply src/test/java/ch/uzh/ifi/hase/soprafs24/controller/UserControllerTest.java:241]: done

**Comment by 5v3nn** [comment]
@Alihan26 done, please look over if changes are ok

**Review by Alihan26 — COMMENTED** [review]
Yes, the changes make sense, from now on the structure will be applied like this @5v3nn

**Review by 5v3nn — APPROVED** [review]
ready to merge
